File: app/src/main/java/com/coderstory/purify/utils/XposedHelper.java
package com.coderstory.purify.utils;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Set;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;

import static de.robv.android.xposed.XposedHelpers.getParameterTypes;

public class XposedHelper {

    public static Class findClass(String classpatch, ClassLoader classLoader) {
        try {
            return XposedHelpers.findClass(classpatch, classLoader);
        } catch (XposedHelpers.ClassNotFoundError error) {
            XposedBridge.log(error.getMessage());
        }
        return null;
    }


    public static void findAndHookMethod(String p1, ClassLoader lpparam, String p2, Object... parameterTypesAndCallback) {
        try {
            XposedHelpers.findAndHookMethod(p1, lpparam, p2, parameterTypesAndCallback);

        } catch (Throwable error) {
            XposedBridge.log(error);
        }
    }

    public static void findAndHookMethod(Class<?> p1, String p2, Object... parameterTypesAndCallback) {
        try {
            XposedHelpers.findAndHookMethod(p1, p2, parameterTypesAndCallback);
        } catch (Throwable error) {
            XposedBridge.log(error);
        }
    }

    public static void hookAllConstructors(String p1,ClassLoader classLoader, XC_MethodHook parameterTypesAndCallback) {
        try {
            Class packageParser = XposedHelpers.findClass(p1, classLoader);
            XposedBridge.hookAllConstructors(packageParser, parameterTypesAndCallback);

        } catch (Throwable error) {
            XposedBridge.log(error);
        }
    }

    protected static void findAndHookMethod(String p1, String p2, Object[] p3) {
        try {
            XposedHelpers.findAndHookMethod(Class.forName(p1), p2, p3);
        } catch (Throwable error) {
            XposedBridge.log(error);
        }
    }

    public static void hookAllMethods(String p1, ClassLoader lpparam, String methodName, XC_MethodHook parameterTypesAndCallback) {
        try {
            Class packageParser = XposedHelpers.findClass(p1, lpparam);
            XposedBridge.hookAllMethods(packageParser, methodName, parameterTypesAndCallback);

        } catch (Throwable error) {
            XposedBridge.log(error);
        }
    }

    public static Set<XC_MethodHook.Unhook> hookAllMethods(Class<?> hookClass, String methodName, XC_MethodHook callback) {
        try {
            return XposedBridge.hookAllMethods(hookClass, methodName, callback);
        } catch (Throwable error) {
            XposedBridge.log(error.getMessage());
        }
        return null;
    }

    public static void writeFile(File file, File file1) {
        FileInputStream fileInputStream;
        FileOutputStream fileOutputStream;
        BufferedInputStream bufferedInputStream;
        BufferedOutputStream bufferedOutputStream;
        try {
            fileInputStream = new FileInputStream(file);
            fileOutputStream = new FileOutputStream(file1);
            if (!file1.getParentFile().exists()) {
                file1.getParentFile().mkdirs();
            }
            bufferedInputStream = new BufferedInputStream(fileInputStream);
            bufferedOutputStream = new BufferedOutputStream(fileOutputStream);
            byte[] bytes = new byte[5120];
            while (true) {
                int read = bufferedInputStream.read(bytes);
                if (read == -1) {
                    break;
                }

                bufferedOutputStream.write(bytes, 0, read);
            }
            bufferedOutputStream.flush();
            bufferedInputStream.close();
            bufferedOutputStream.close();
            fileOutputStream.close();
            fileInputStream.close();

        } catch (IOException error) {
            XposedBridge.log(error);
        }
    }

    /**
     * 调用指定的方法并返回结果
     *
     * @param targetObject 方法所在对象
     * @param methodName   方法的名字
     * @param returnType   返回类型
     * @param params       参数类型
     * @return
     * @throws InvocationTargetException
     * @throws IllegalAccessException
     */
    public static Object findResultByMethodNameAndReturnTypeAndParams(Object targetObject, String methodName, String returnType, Object... params) throws InvocationTargetException, IllegalAccessException {
        return findMethodByNameAndReturnType(targetObject.getClass(), methodName, returnType, getParameterTypes(params)).invoke(targetObject, params);
    }

    public static Method findMethodByNameAndReturnType(Class<?> targetObject, String methodName, String returnType, Class<?>... params) {
        for (Method method : targetObject.getDeclaredMethods()) {
            if (method.getReturnType().getName().equals(returnType) && method.getName().equals(methodName)) {
                Class[] parameterTypes = method.getParameterTypes();
                if (params.length != parameterTypes.length) {
                    continue;
                }
                for (int i = 0; i < params.length; i++) {
                    if (params[i] != parameterTypes[i]) {
                        break;
                    }
                }
                method.setAccessible(true);
                return method;
            }
        }
        throw new NoSuchMethodError();
    }

    public static Field findFieldByClassAndTypeAndName(Class<?> targetObject, Class<?> fieldType, String fieldName) {
        Class clazz = targetObject;
        do {
            for (Field field : clazz.getDeclaredFields()) {
                if (field.getType() == fieldType && field.getName().equals(fieldName)) {
                    field.setAccessible(true);
                    return field;
                }
            }
            clazz = clazz.getSuperclass();
        } while (clazz != null);
        throw new NoSuchFieldError("Field of type " + fieldType.getName() + " in class " + targetObject.getName());
    }

    protected void hookAllConstructors(Class<?> hookClass, XC_MethodHook callback) {
        try {
            XposedBridge.hookAllConstructors(hookClass, callback);
        } catch (Throwable error) {
            XposedBridge.log(error.getMessage());
        }
    }

    protected Class findClassWithoutLog(String classpatch, ClassLoader classLoader) {
        try {
            return XposedHelpers.findClass(classpatch, classLoader);
        } catch (XposedHelpers.ClassNotFoundError error) {
            error.printStackTrace();
        }
        return null;
    }

}

File: app/src/main/java/com/coderstory/flyme10/tools/XposedHelper.kt
package com.coderstory.flyme10.tools

import com.coderstory.flyme10.BuildConfig
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XSharedPreferences
import de.robv.android.xposed.XposedBridge
import de.robv.android.xposed.XposedHelpers
import de.robv.android.xposed.XposedHelpers.ClassNotFoundError
import java.io.File

open class XposedHelper {
    var prefs: XSharedPreferences
    fun hookAllConstructors(p1: String, parameterTypesAndCallback: XC_MethodHook) {
        try {
            val packageParser = findClass(p1, null)
            hookAllConstructors(packageParser, parameterTypesAndCallback)
        } catch (e: Throwable) {
            if (BuildConfig.DEBUG) XposedBridge.log(e)
        }
    }

    protected fun hookAllConstructors(
        hookClass: Class<*>,
        callback: XC_MethodHook
    ): Set<XC_MethodHook.Unhook> {
        return try {
            val result = XposedBridge.hookAllConstructors(hookClass, callback)
            if (result.size == 0) {
                XposedBridge.log("类" + hookClass.name + "中的构造方法没有被hook到")
            }
            result
        } catch (error: Throwable) {
            XposedBridge.log(error)
            HashSet()
        }
    }

    protected fun findClassWithoutLog(classpatch: String, classLoader: ClassLoader): Class<*>? {
        return try {
            XposedHelpers.findClass(classpatch, classLoader)
        } catch (error: Exception) {
            null
        }
    }


    companion object {
        fun findClass(classpatch: String, classLoader: ClassLoader?): Class<*> {
            try {
                return XposedHelpers.findClass(classpatch, classLoader)
            } catch (error: ClassNotFoundError) {
                XposedBridge.log(error)
            }
            return String::class.java
        }

        fun findAndHookMethod(
            p1: String,
            param: ClassLoader,
            p2: String,
            vararg parameterTypesAndCallback: Any?
        ) {
            try {
                XposedHelpers.findAndHookMethod(p1, param, p2, *parameterTypesAndCallback)
            } catch (error: Throwable) {
                XposedBridge.log(error)
            }
        }

        fun findAndHookMethod(p1: Class<*>, p2: String, vararg parameterTypesAndCallback: Any) {
            try {
                XposedHelpers.findAndHookMethod(p1, p2, *parameterTypesAndCallback)
            } catch (error: Throwable) {
                XposedBridge.log(error)
            }
        }

        fun hookAllConstructors(
            p1: String,
            classLoader: ClassLoader,
            parameterTypesAndCallback: XC_MethodHook
        ) {
            try {
                val packageParser = XposedHelpers.findClass(p1, classLoader)
                XposedBridge.hookAllConstructors(packageParser, parameterTypesAndCallback)
            } catch (error: Throwable) {
                XposedBridge.log(error)
            }
        }

        protected fun findAndHookMethod(p1: String, p2: String, p3: Array<Any>) {
            try {
                XposedHelpers.findAndHookMethod(Class.forName(p1), p2, *p3)
            } catch (error: Throwable) {
                XposedBridge.log(error)
            }
        }

        fun hookAllMethods(
            p1: String,
            classLoader: ClassLoader?,
            methodName: String,
            parameterTypesAndCallback: XC_MethodHook
        ): Int {
            return try {
                val packageParser = XposedHelpers.findClass(p1, classLoader)
                val count = XposedBridge.hookAllMethods(
                    packageParser,
                    methodName,
                    parameterTypesAndCallback
                ).size
                if (count == 0) {
                    XposedBridge.log("类" + p1 + "中的方法" + methodName + "没有被hook到")
                }
                count
            } catch (error: Throwable) {
                XposedBridge.log(error)
                0
            }
        }

        fun hookAllMethods(
            hookClass: Class<*>,
            methodName: String,
            callback: XC_MethodHook
        ): Set<XC_MethodHook.Unhook>? {
            try {
                val result = XposedBridge.hookAllMethods(hookClass, methodName, callback)
                if (result.size == 0) {
                    XposedBridge.log("类" + hookClass.name + "中的方法" + methodName + "没有被hook到")
                }
                return result
            } catch (error: Throwable) {
                XposedBridge.log(error)
            }
            return null
        }
    }

    init {
        prefs = XSharedPreferences(BuildConfig.APPLICATION_ID, Misc.SharedPreferencesName)
        if (prefs.all.keys.size == 0) {
            prefs =
                XSharedPreferences(File("/data/user_de/0/" + Misc.ApplicationName + "/shared_prefs/" + Misc.SharedPreferencesName + ".xml"))
        }
    }
}